assets/js/base/context/checkout-context.js

```diff
@@ -0,0 +1,59 @@
+/**
+ * External dependencies
+ */
+import {
+	createContext,
+	useContext,
+	useState,
+	useMemo,
+} from '@wordpress/element';
+
+const CheckoutContext = createContext( {} );
+
+export const useCheckoutContext = () => {
+	return useContext( CheckoutContext );
+};
+
+const CheckoutProvider = ( { children, initialActivePaymentMethod } ) => {
+	const [ successRedirectUrl, setSuccessRedirectUrl ] = useState( '' );
+	const [ failureRedirectUrl, setFailureRedirectUrl ] = useState( '' );
+	const [ checkoutComplete, setCheckoutComplete ] = useState( false );
+	const [ checkoutHasError, setCheckoutHasError ] = useState( false );
+	const [ notices, updateNotices ] = useState( [] );
+	const [ isCalculating, setIsCalculating ] = useState( false );
+	const [ activePaymentMethod, setActivePaymentMethod ] = useState(
+		initialActivePaymentMethod
+	);
+	const contextValue = useMemo( () => {
```

Comment:
My understanding of context may be flawed.  I was concerned since I'm using an object for the context `value` that anytime that object is changed, it invalidates the tree inside the context provider causing re-renders.  So this was an attempt to limit that re-rendering only if absolutely necessary (so don't re-create the object unless internal property values change).

If my understanding is flawed please correct me :) Otherwise, I probably should validate that assumption in a test!

Reply:
🤔 You're correct. Hm... if the object _properties_ change, then we definitely want to re-render anyway... but if the object _itself_ changes, which it will do on every render of this component, it may also cause a re-render, which we may _not_ want. So maybe _I've_ learned something! I should also validate this assumption.

Reply:
So I'm basing my assumption on this paragraph from the React docs (still should be validated, because I know docs can be wrong sometimes!):

> All consumers that are descendants of a Provider will re-render whenever the Provider’s value prop changes. The propagation from Provider to its descendant consumers is not subject to the shouldComponentUpdate method, so the consumer is updated even when an ancestor component bails out of the update.